### PR TITLE
Ensure that edge‑case browsers are only allowed in specific categories

### PR DESCRIPTION
**Pull request by ExE-Boss**
This has been suggested by @connorshea in https://github.com/mdn/browser-compat-data/pull/2332#issuecomment-405315866, and I agree that it’s a good idea, as it ensures that contributors don’t try to add `firefox_ios` or `thunderbird` to non‑WebExtension entries, or `nodejs` to non‑JavaScript entries (which it turns out they already did).

It also ensures that PRs like #2484 get better error messages.

---

review?(@a2sheppy, @Elchi3)

### CODE REVIEW

File: test/test-browsers.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  'webextensions-desktop': [
+    'chrome',
+    'edge',
+    'firefox',
+    'opera',
+    'thunderbird',
+  ],
+  'webextensions-mobile': [
+    'firefox_android',
+    'firefox_ios',
```
Comment (evilpie) [inline test/test-browsers.js]: I think only Firefox for Android has webextensions

Comment (ExE-Boss) [reply test/test-browsers.js]: Well, Firefox for iOS hasn’t yet been added to the schema, so it will fail validation regardless.

The reason I already added it and Thunderbird here is so that I don’t have to rebase #2332 and #2333 or this to add them once they’re merged.

**Comment by Elchi3** [comment]
I think this is no longer relevant now. In https://github.com/mdn/browser-compat-data/pull/3160 we've seen that node implements some features in api/ land and so that validation isn't really useful anymore. For the other edge case browsers, we have decided to not add them to the repo. 

I could imagine to test for webextension-only browsers, but I'm not sure if it is really neeeded. What do you think, @ExE-Boss ?

**Comment by ExE-Boss** [comment]
I think it might be useful.

**Comment by ExE-Boss** [comment]
The way I understand it, the goal for MDN is to only display Node in the `api` category when it’s present, so I removed the `nodejs` keys from HTML, Audio and Media entries.

File: test/test-browsers.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  const category = (/^(?:\.[\/\\])?([^\/\\]+)/.exec(relativePath) || [])[1];
+  const data = require(filename);
+
+  if (!category) {
+    console.log('\x1b[1;30m  Browsers – Unknown \x1b[0m');
+    return false;
+  }
+
+  let bcCategory = 'web';
+  let displayBrowers = [...browsers['desktop'], ...browsers['mobile']];
```
[the remark this answers is not in the document]
Comment (ExE-Boss) [reply test/test-browsers.js]: That’s from the `{{Compat}}` macro: https://github.com/mdn/kumascript/blob/6550757cc685355532d6a6a468825f32d5cae5b0/macros/Compat.ejs#L57

File: test/test-browsers.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  for (const key in data) {
+    if (key === "__compat") continue;
+    hasErrors |= processData(data[key], browsers, category, logger, (path && path.length > 0) ? `${path}.${key}` : key);
+  }
+  return hasErrors;
+}
+
+function testBrowsers(filename) {
+  const relativePath = path.relative(path.resolve(__dirname, '..'), filename);
+  const category = (/^(?:\.[\/\\])?([^\/\\]+)/.exec(relativePath) || [])[1];
```
[the remark this answers is not in the document]
Comment (ExE-Boss) [reply test/test-browsers.js]: Because Windows.

Although, I should probably use the&nbsp;`path` module instead.

File: test/test-browsers.js (hunk 1 above)
Comment (a2sheppy) [reply test/test-browsers.js]: Firefox for iOS will have WebExtensions eventually; it's on their roadmap for this year as I understand it.

File: test/test-browsers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Elchi3) [reply test/test-browsers.js]: Yes, this is better, thanks! One more question: Why is there `|| [])`?

File: test/test-browsers.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Elchi3) [reply test/test-browsers.js]: Hah, bugs everywhere :) Thanks for fixing it here now.

File: test/test-browsers.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    'firefox_android',
+  ],
+};
+
+function processData(data, browsers, category, logger, path = '') {
+  let hasErrors = false;
+  if (data.__compat && data.__compat.support) {
+    const invalidEntries = Object.keys(data.__compat.support).filter(value => !browsers.includes(value));
+    if (invalidEntries.length > 0) {
+      logger.error(`'${path}' has the following browesers, which are invalid for ${category}: ${invalidEntries.join(', ')}`);
```
[the remark this answers is not in the document]
Comment (Elchi3) [reply test/test-browsers.js]: The typo "browesers" is still there.

File: test/test-browsers.js (hunk 1 above)
Comment (Elchi3) [reply test/test-browsers.js]: As said in #2332 "We can re-consider this [Firefox iOS] if it should become a reality."

File: test/test-browsers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ExE-Boss) [reply test/test-browsers.js]: That was in&nbsp;case the&nbsp;RegExp didn’t&nbsp;match for&nbsp;whatever reason.
